File: src/lib/services/payment.ts
import { SupabaseClient } from '@supabase/supabase-js';
import { ACQUIRING_API_KEY, ACQUIRING_BASE_URL } from '../config';

export class PaymentService {
  constructor(private db: SupabaseClient) {}

  async charge(
    userId: string,
    amount: number,
    sessionId?: string | null,
  ): Promise<Record<string, unknown>> {
    const { data: user } = await this.db.from('users').select('*').eq('id', userId).maybeSingle();
    if (!user) throw new Error('User not found');

    const { data: card } = await this.db
      .from('payment_cards')
      .select('*')
      .eq('user_id', userId)
      .eq('is_active', true)
      .maybeSingle();

    const { data: payment, error } = await this.db
      .from('payments')
      .insert({
        user_id: userId,
        session_id: sessionId || null,
        amount,
        status: 'PENDING',
        card_token: card?.card_token || null,
      })
      .select()
      .single();
    if (error) throw new Error(error.message);

    if (!card) return this.handleInsufficientFunds(sessionId, amount, userId, payment.id);

    let success = false;
    try {
      success = await this.callAcquiring(card.card_token, amount, payment.id);
    } catch {
      success = false;
    }

    if (success) {
      await this.db.from('payments').update({ status: 'SUCCESS' }).eq('id', payment.id);
      if (sessionId) {
        await this.db.from('sessions').update({ is_paid: true }).eq('id', sessionId);
      }
      return { status: 'success', payment_id: payment.id };
    }
    return this.handleInsufficientFunds(sessionId, amount, userId, payment.id);
  }

  private async callAcquiring(
    cardToken: string,
    amount: number,
    paymentId: string,
  ): Promise<boolean> {
    try {
      const resp = await fetch(`${ACQUIRING_BASE_URL}/charge`, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${ACQUIRING_API_KEY}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ card_token: cardToken, amount, order_id: paymentId }),
        signal: AbortSignal.timeout(30000),
      });
      if (!resp.ok) return false;
      const data = await resp.json();
      return data.status === 'success';
    } catch {
      return false;
    }
  }

  private async handleInsufficientFunds(
    _sessionId: string | null | undefined,
    amount: number,
    userId: string,
    paymentId: string,
  ): Promise<Record<string, unknown>> {
    await this.db.from('payments').update({ status: 'FAILED' }).eq('id', paymentId);
    const { data: user } = await this.db.from('users').select('debt_amount').eq('id', userId).maybeSingle();
    const newDebt = (user?.debt_amount ?? 0) + amount;
    await this.db
      .from('users')
      .update({ has_debt: true, debt_amount: newDebt })
      .eq('id', userId);
    return { status: 'debt', debt_amount: newDebt };
  }

  async processDebt(userId: string): Promise<boolean> {
    const { data: user } = await this.db.from('users').select('*').eq('id', userId).maybeSingle();
    if (!user || !user.has_debt || user.debt_amount <= 0) return true;

    const { data: card } = await this.db
      .from('payment_cards')
      .select('*')
      .eq('user_id', userId)
      .eq('is_active', true)
      .maybeSingle();
    if (!card) return false;

    const { data: payment, error } = await this.db
      .from('payments')
      .insert({
        user_id: userId,
        amount: user.debt_amount,
        status: 'PENDING',
        card_token: card.card_token,
      })
      .select()
      .single();
    if (error) return false;

    const success = await this.callAcquiring(card.card_token, user.debt_amount, payment.id);
    if (success) {
      await this.db.from('payments').update({ status: 'SUCCESS' }).eq('id', payment.id);
      await this.db
        .from('users')
        .update({ has_debt: false, debt_amount: 0 })
        .eq('id', userId);
      return true;
    }
    await this.db.from('payments').update({ status: 'FAILED' }).eq('id', payment.id);
    return false;
  }
}

File: backend/app/services/payment.py
import httpx
import logging
from app.config import ACQUIRING_API_KEY, ACQUIRING_BASE_URL

logger = logging.getLogger(__name__)


class PaymentService:
    def __init__(self, db):
        self.db = db

    async def charge(self, user_id: str, amount: float, session_id: str | None = None) -> dict:
        user = await self.db.user.find_unique(where={"id": user_id})
        if not user:
            raise ValueError("User not found")

        card = await self.db.paymentcard.find_first(
            where={"userId": user_id, "isActive": True}
        )

        payment = await self.db.payment.create(data={
            "userId": user_id,
            "sessionId": session_id,
            "amount": amount,
            "status": "PENDING",
            "cardToken": card.cardToken if card else None,
        })

        if not card:
            return await self.handle_insufficient_funds(session_id, amount, user_id, payment.id)

        try:
            success = await self._call_acquiring(card.cardToken, amount, payment.id)
        except Exception as e:
            logger.error(f"Acquiring error: {e}")
            success = False

        if success:
            await self.db.payment.update(where={"id": payment.id}, data={"status": "SUCCESS"})
            if session_id:
                await self.db.session.update(where={"id": session_id}, data={"isPaid": True})
            return {"status": "success", "payment_id": payment.id}
        else:
            return await self.handle_insufficient_funds(session_id, amount, user_id, payment.id)

    async def _call_acquiring(self, card_token: str, amount: float, payment_id: str) -> bool:
        async with httpx.AsyncClient() as client:
            try:
                resp = await client.post(
                    f"{ACQUIRING_BASE_URL}/charge",
                    headers={"Authorization": f"Bearer {ACQUIRING_API_KEY}"},
                    json={"card_token": card_token, "amount": amount, "order_id": payment_id},
                    timeout=30,
                )
                return resp.status_code == 200 and resp.json().get("status") == "success"
            except Exception:
                return False

    async def handle_insufficient_funds(
        self, session_id: str | None, amount: float, user_id: str, payment_id: str
    ) -> dict:
        await self.db.payment.update(where={"id": payment_id}, data={"status": "FAILED"})
        # Mark debt — do NOT cut power
        user = await self.db.user.find_unique(where={"id": user_id})
        new_debt = (user.debtAmount or 0) + amount
        await self.db.user.update(
            where={"id": user_id},
            data={"hasDebt": True, "debtAmount": new_debt},
        )
        logger.warning(f"Debt marked for user {user_id}: {new_debt}")
        return {"status": "debt", "debt_amount": new_debt}

    async def process_debt(self, user_id: str) -> bool:
        user = await self.db.user.find_unique(where={"id": user_id})
        if not user or not user.hasDebt or user.debtAmount <= 0:
            return True

        card = await self.db.paymentcard.find_first(
            where={"userId": user_id, "isActive": True}
        )
        if not card:
            return False

        payment = await self.db.payment.create(data={
            "userId": user_id,
            "amount": user.debtAmount,
            "status": "PENDING",
            "cardToken": card.cardToken,
        })

        success = await self._call_acquiring(card.cardToken, user.debtAmount, payment.id)
        if success:
            await self.db.payment.update(where={"id": payment.id}, data={"status": "SUCCESS"})
            await self.db.user.update(
                where={"id": user_id},
                data={"hasDebt": False, "debtAmount": 0},
            )
            return True
        else:
            await self.db.payment.update(where={"id": payment.id}, data={"status": "FAILED"})
            return False
